### PR TITLE
F4BY Flight Controller MCU upgrade

### PR DESCRIPTION
**Pull request by kozinalexey**
hwdef:F4BY Flight Controller MCU upgrade
replace old MCU STM32F407VGT (1MB Flash) with STM32F427VET rev3 or above (2MB Flash)
fully tested including flight tests with Arducopter 4.5 stable
changed New style Pixracer LED configuration for master repo

### CODE REVIEW

**Comment by Hwurzburg** [comment]
will both boards be available? its wiki page is very dated and has broken links... needs updating if this board is really available

**Comment by kozinalexey** [comment]
hi, add board id [Tools/AP_Bootloader/board_types.txt] is required in this PR ?
i have test chibios / build (configure-all, chibios, 10) (pull_request) Failing

File: libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat [hunk 1]
```diff
@@ -0,0 +1,169 @@
+# hw definition file for processing by chibios_hwdef.py
+# for F4BY v2.1.5 board description http://swift-flyer.com/?page_id=83
+
+# MCU class and specific type
+MCU STM32F4xx STM32F427xx
+
+# board ID for firmware load
+APJ_BOARD_ID 21
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat:8]: Need to reserve an id in Tools/AP_Bootloader/board_list.txt and use its name in here

**Comment by peterbarker** [comment]
> The wiki and the store are under restoration, we decided to continue working on the project

Fantastic!

Something H743 would be nice at some stage :-)

**Comment by kozinalexey** [comment]
> > The wiki and the store are under restoration, we decided to continue working on the project
> 
> Fantastic!
> 
> Something H743 would be nice at some stage :-)
Hello Peter, you are absolutely right, we are already testing the board with H743 for new production. The configuration with 427 MCU is intended mainly for those who have already purchased F4BY and know how to solder. Unfortunately, F4 and H7 in a 100-pin package are not pin-to-pin compatible

File: libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat (hunk 1 above)
Comment (kozinalexey) [reply libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat:8]: ID is reserved. Tools/AP_Bootloader/board_types.txt How can I restart the tests now?

**Comment by kozinalexey** [comment]
> will both boards be available?

Yes, we will provide technical support for both boards. Another one with MCU H743 will appear soon, the first samples are already soldered at the factory

**Comment by kozinalexey** [comment]
"Some checks haven’t completed yet
47 queued and 2 successful checks "
Last night all tests were completed in 10 minutes

**Comment by kozinalexey** [comment]
image
i selected ID 21 could this be causing the test error?

**Comment by Hwurzburg** [comment]
you did not include the bootloader with the new ID...
@peterbarker will have to comment on using the low ID number

File: libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+PE11 TIM1_CH2 TIM1 PWM(6) GPIO(55)
+PE13 TIM1_CH3 TIM1 PWM(7) GPIO(56)
+PE14 TIM1_CH4 TIM1 PWM(8) GPIO(57)
+
+PD13 TIM4_CH2 TIM4 PWM(9) GPIO(58)
+PD12 TIM4_CH1 TIM4 PWM(10) GPIO(59)
+PD15 TIM4_CH4 TIM4 PWM(11) GPIO(60)
+PD14 TIM4_CH3 TIM4 PWM(12) GPIO(61)
+
+# also USART6_RX for serial RC
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat:131]:
```suggestion
# also USART6_RX for unidirectional RC,including PPM
```

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
-move all default params into hwdef
- setup SD card in bootloader to enable SD card firmware upgrade

**Comment by kozinalexey** [comment]
> * setup SD card in bootloader to enable SD card firmware upgrade

I have an almost ready board with an h743 MCU in it, I will make a bootloader from SD since the H7 bootloader size is 128. With f427 it is important to include it in the release faster, the CAN, Gimbals, Camera functions are in great demand by users

**Comment by Hwurzburg** [comment]
Uses flash unnecessarily

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat [hunk 3]
```diff
@@ -0,0 +1,169 @@
+# hw definition file for processing by chibios_hwdef.py
+# for F4BY v2.1.5 board description http://swift-flyer.com/?page_id=83
+
+# MCU class and specific type
+MCU STM32F4xx STM32F427xx
+
+# board ID for firmware load
+APJ_BOARD_ID 7221
```
[the remark this answers is not in the document]
Comment (kozinalexey) [reply libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat:8]: I have already sent the compiled bin with id 7221, will it be rebuilt automatically or will it have to be rebuilt and updated manually?

**Comment by kozinalexey** [comment]
excellent, I myself didn’t think of using a define in board_types.txt instead of specifying the ID in three places

**Comment by peterbarker** [comment]
Commit list needs cleaning up.  I can do that if everyone's happy with the PR otherwise.

We're after two commits:
```
board_types.txt: add new board type for F4BY_F427
```
and
```
hwdef: add F4BY_F427
```

**Comment by kozinalexey** [comment]
> Commit list needs cleaning up. I can do that if everyone's happy with the PR otherwise.
> 
> We're after two commits:
> 
> ```
> board_types.txt: add new board type for F4BY_F427
> ```
> 
> and
> 
> ```
> hwdef: a
> ```
Hi, done. All tests passed

**Comment by Hwurzburg** [comment]
you pushed again after saying its done....is it really done?, if no more pushes I will review...

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
you haver not removed the defaults.param file...all those should be moved into the hwdef.dat file

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
you have not removed the defaults.param file and moved them into the hwdef.dat file

File: libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/defaults.parm [hunk 4]
```diff
@@ -0,0 +1,7 @@
+BATT_AMP_OFFSET 2.316
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/defaults.parm:1]: Replace this with a line like this in your hwdef:
```
define AP_BATT_CURR_AMP_OFFSET_DEFAULT 2.316
```

See `AP_BattMonitor_Analog.cpp` for the other default-define names.

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Comment by kozinalexey** [comment]
> you have not removed the defaults.param file and moved them into the hwdef.dat file

thanks, ok, I'll do it. I couldn't find the instructions.

File: libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/defaults.parm (hunk 4 above)
Comment (kozinalexey) [reply libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/defaults.parm:1]: > See `AP_BattMonitor_Analog.cpp` for the other default-define names.

Thank you, Peter

**Comment by kozinalexey** [comment]
is it correct style? 
found in another board setup:
https://github.com/ArduPilot/ardupilot/blob/ec1dc7cbc558ba94e687344bb084ac6fef113216/libraries/AP_HAL_ChibiOS/hwdef/Durandal/hwdef.dat#L326C1-L332C34

# battery setup
define HAL_BATT_MONITOR_DEFAULT 4
define HAL_BATT_VOLT_PIN 13
define HAL_BATT_CURR_PIN 12
define HAL_BATT_VOLT_SCALE 16.04981
define HAL_BATT_CURR_SCALE 100

i found only one _default in AP_BattMonitor_Analog.cpp : 
define **HAL_BATT_MONITOR_DEFAULT** 4


Peter says 
define AP_BATT_CURR_AMP_OFFSET_DEFAULT 2.316

so:
define HAL_BATT_MONITOR_DEFAULT 4

define AP_BATT_VOLT_PIN_DEFAULT 13
define AP_BATT_CURR_PIN_DEFAULT 12
define AP_BATT_CURR_AMP_OFFSET_DEFAULT 2.316
define AP_BATT_AMP_PERVLT_DEFAULT 100
define AP_BATT_VOLT_MULT_DEFAULT 16.04981
is correct?

**Comment by kozinalexey** [comment]
with these parameters everything works correctly
```
# this constants with HAL_
define HAL_BATT_MONITOR_DEFAULT 4
define HAL_BATT_VOLT_PIN 13
define HAL_BATT_CURR_PIN 12
define HAL_BATT_VOLT_SCALE 16.04981
define HAL_BATT_CURR_SCALE 100

# this constant with AP_
define AP_BATT_CURR_AMP_OFFSET_DEFAULT 2.316
```

File: libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+# this constants with HAL_
+define HAL_BATT_MONITOR_DEFAULT 4
+define HAL_BATT_VOLT_PIN 13
+define HAL_BATT_CURR_PIN 12
+define HAL_BATT_VOLT_SCALE 16.04981
+define HAL_BATT_CURR_SCALE 100
+
+# this constant with AP_
+define AP_BATT_CURR_AMP_OFFSET_DEFAULT 2.316
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat:177]: wow! that is a huge offset! most need none or something very small.....this is suspicious

Comment (kozinalexey) [reply libraries/AP_HAL_ChibiOS/hwdef/F4BY_F427/hwdef.dat:177]: > wow! that is a huge offset! most need none or something very small.....this is suspicious

We use a bipolar sensor based on a hall effect Allegro ACS758. It measures both discharge and charge. thus its midpoint is about half the voltage

**Comment by kozinalexey** [comment]
> wiki page is very dated and has broken links... needs updating

documentation done.
updated links in README.md

File: Tools/AP_Bootloader/board_types.txt [hunk 6]
```diff
@@ -416,7 +416,7 @@ AP_HW_CUAV-7-NANO                    7000
 AP_HW_VUAV-V7pro                     7100
 
 # please fill gaps in the above ranges rather than adding past ID #7109
-
+AP_HW_F4BY_F427                        1530
```
Comment (Hwurzburg) [inline Tools/AP_Bootloader/board_types.txt:419]: move up in file so as to be in proper order of numbers

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
need to change board_types file
need to split into two commits....one for Tools library and one for AP_HAL_ChibiOS library

I tried to do this for you but your PR does not allow changes by other devs

File: Tools/AP_Bootloader/board_types.txt (hunk 6 above)
Comment (kozinalexey) [reply Tools/AP_Bootloader/board_types.txt:429]: > move up in file so as to be in proper order of numbers

I ran rebase to my path, now I need to wait again for the tests to finish. we take too long, this is causing problems.
It’s good that no one has taken my ID yet.

Comment (kozinalexey) [reply Tools/AP_Bootloader/board_types.txt]: > move up in file so as to be in proper order of numbers

done, moved from line 420 to 335

**Comment by kozinalexey** [comment]
> move up in file so as to be in proper order of numbers
done moved from 420 to 335 line

for prepare code to merge i must to run checks. Checks need actual version of master. When i press rebase i see lost comments.

**Comment by kozinalexey** [comment]
> need to split into two commits....one for Tools library and one for AP_HAL_ChibiOS library
only two comits in one PR?
or move second commit to another PR?

**Comment by kozinalexey** [comment]
> need to split into two commits....one for Tools library and one for AP_HAL_ChibiOS library

done

**Comment by Hwurzburg** [comment]
sorry the commit format is still wrong...would be an easy fix for me or @peterbarker  if we had access permission to your branch

as to board id, you could have created a PR just for the board id bt itself and it would be merged in a day or two since its just a reservation.....

**Comment by peterbarker** [comment]
Merged this via the other PR I created as I couldn't force-push this branch.

**Comment by kozinalexey** [comment]
> Merged this via the other PR I created as I couldn't force-push this branch.

Thank you so much, great job!

**Pull request closed without merging**